LSPS1 - Channel Request

SeverinAlexB opened the pull request:
This is a proposal to update the `LSPS1 channel request` spec to fix the issues we discussed in the [2023-02-1 meeting](https://github.com/BitcoinAndLightningLayerSpecs/meetings/blob/main/2023-02-01.md) and the discussion in the Telegram group after.

The goal of this proposal is to
- Improve modularity of the spec so interested parties can choose what part of the spec they like to implement.
- Make request/response values more clear so implementing parties end up with the same api.
- Generally add more descriptions so external parties have a chance to understand the spec.

## Content

The proposed api consists of a base_api that is required to be implemented by all parties and optional extensions. Extensions are additional fields / endpoints that support a certain functionality.

- `base_api` Channel purchase and open.
- `lnurl_lud2` LNURL channel open.
- `jit_channels` Just-In-Time channels.
- `refunds` Automatic refunds (newly proposed by me).

I am happy to split this PR up if demanded but I think it's worth having this full PR so everybody gets an initial impression of my proposal.

### OpenApi

The proposal leans heavily on the [OpenApi format](https://en.wikipedia.org/wiki/OpenAPI_Specification) to describe request / response values. Use any OpenApi editor like the [Swagger Editor](https://editor.swagger.io/) to parse the content. All OpenApi docs in this pull request include descriptions for every field. So please have a look at them.

Thanks for your feedback.

JssDWt commented on LSPS1/extensions/jit_channels/LSPS1.jit_channels.yaml:
```diff
@@ -0,0 +1,115 @@
+        properties:
+          node_connection_string:
+            description: "The node connection string that the channel will be opened to in case of an incoming payment."
+            example: "03864ef025fde8fb587d989186ce6a4a186895ee44a926bfc370e2c366597a3f8f@3.33.236.230:9735"
```
One thing about JIT channels is for mobile nodes is the IP address changes often. The client would then have to connect its lightning node to the LSP in order for the LSP to open a channel to its pubkey. I can't think of a way to describe that other than the readme. Maybe readme is fine.

SeverinAlexB replied:
You need to wake up the node anyway for the channel open, don't you? So connecting is not an issue. Readme should be enough. Should we transform this property from `node_connection_string` to `node_connection_string_or_pubkey`?

ZmnSCPxj-jr replied:
Might also want to be explicit about exactly what the format is for `lsp_connection_string` and `node_connection_string`.  For example, Core Lightning supports the grammar `<node> '@' '[' <ip> ']' ':' <port>`, e.g. `023456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef012@[::1]:42`.  The reason for this is that colons are used in IPv6 addresses, and Core Lightning supports leaving out the port number, i.e. `<node> '@' <ip>` since the BOLT spec specifies you should always use 9735 anyway.

```markdown
The `lsp_connection_string` MUST match one of the following grammars:

* `<node> '@' <ip>`, without any `:` characters inside `<ip>`; this implies the default port for the network, 9735 for Bitcoin Lightning, 19735 for Testnet Bitcoin Lightning.
* `<node> '@' <ip> `:` <port>`, with the last `:` after `@` considered a separator for the port number.
* `<node> '@' '[' <ip> ']'`, with any `:` inside `<ip>` considered part of the IP address, and using the default port for the network; this is needed for IPv6 addresses.
* `<node> '@' '[' <ip> ']' ':' <port>`, with any `:` inside `<ip>` considered part of the IP address.
```

SeverinAlexB replied:
I will turn this into a general node connection string definition in the spec (for `lsp_connection_string` and `user_connection_string`).

I would like to remove the lines `<node> '@' <ip>` and `<node> '@' '[' <ip> ']'` that implicitly assume a port. This would simplify the spec and make the connection string more explicit.

I would also like to add support for TOR v3 addresses:
* `<node> '@' tor_onion_v3_address ':' <port>`, with the last `:` after `@` considered a separator for the port number.

> Rationale for only supporting v3: v2 addresses are [deprecated](https://support.torproject.org/onionservices/v2-deprecation/) since July 2021.

SeverinAlexB replied:
The user should be allowed to provide a connection_string OR a pubkey. Mobile wallets usually don't have a public IP so they are not able to provide a full connection_string anyway.

I want to change the spec so the peer connection MUST be established by the user and remove the LSP connection requirement completely. My gut feeling tells me that we should add "LSP MAY open a peer connection" though.

SeverinAlexB replied:
>Well, they still might try, and this connection string might be bogus. So it should be clear from the spec that these are really two entirely separate paths and you MUST NOT provide a full bogus connection string.

I am not sure I understand why separate paths? In theory, if the user always establishes the peer connection the user_pubkey is enough. The user_connection_string is only necessary in case the LSP wants to establish the connection. Maybe we should fully drop the user_connection_string and only use a user_pubkey instead?

> Yes, I think this makes a lot of sense at least for any NATed case. For any other case it of course might result in some unnecessary delay, as the client can only establish the connection after getting the response, while the LSP could already establish the connection in parallel to sending the response. It's probably negligible though?

In theory, the user can establish a peer connection before they attempt the payment. This way, the channel can be opened instantly as soon as the payment arrived. This should be even faster than establishing the peer connection after the payment arrived.

JssDWt commented on LSPS1/extensions/jit_channels/LSPS1.jit_channels.yaml:
```diff
@@ -0,0 +1,115 @@
+                description: "The node_id the user puts in the route hint of the invoice."
+                example: "031015a7839468a3c266d662d5bb21ea4cea24226936e2864a7ca4f2c3939836e0"
+                type: "string"
+          registerd_node_connection_string:
```
nit: `registered`

SeverinAlexB replied:
ef794f2

JssDWt commented on LSPS1/extensions/refunds/LSPS1.refunds.yaml:
```diff
@@ -0,0 +1,154 @@
+
+
+paths:           
+  /lsp/channels/{id}/refund:
```
Regarding privacy, would it be better to put the channel id in the request itself? Someone sniffing could see the channel id and request the information itself, because there is no form of authentication (I imagine authentication would be hard to implement too). If the parameters are in the request body, and transport security is used, at least only the LSP is able to read the channel information.
That would break the REST setup of the proposal though. It would become more of an RPC style spec with POST requests for everything.

SeverinAlexB replied:
Good point especially as the id kind of acts as a secret for refunds.

SeverinAlexB replied:
Are you sure though? https://stackoverflow.com/a/499594

SeverinAlexB replied:
Closing this discussion as I think privacy is given (see stackoverflow) and I didn't get any more inputs from you @JssDWt.

JssDWt reviewed (commented):
I'm a fan of the openapi spec. It's great that developers implementing clients can generate client code from the spec. The extension model also looks promising and extensible (indeed). 

I left somme comments in places I'm unsure about. I will generate the client code next week, see what it looks like.

JssDWt commented on LSPS1/LSPS1.yaml:
```diff
@@ -0,0 +1,430 @@
+            description: "Inbound liquidity amount in satoshis. Funds are at the LSP side."
+            example: 5000000
+            minimum: 0
+          local_balance:
```
Thinking about this a little more. An option could be that every endpoint's requests and responses are constructed with envelope object, where extensions are able to add fields to. So `ChannelOrderRequest` could become:
```
{
    "version": 1 // spec version
    "channel_order_request": {
       // the actual request as defined by the base spec
    },
    "local_balance": { 
        "version": 12,
        // values for the local balance extension version 12 go here
    }
}
```

SeverinAlexB replied:
### API Key
All things that are about api-keys, discounts or any custom modification of the order, I added the `coupon_code` for. Originally, I added this field for channel sales, in case you want to give the customer 10% off or so. Nothing speaks against using it with an apiKey IMO. 

### Request/Response structure
That's an interesting approach. I am a bit worried we over engineer the api though.

SeverinAlexB commented on LSPS1/LSPS1.yaml:
```diff
@@ -0,0 +1,430 @@
+          extensions:
+            type: array
+            items:
+              $ref: "#/components/schemas/Extension"
```
Can definitely do that!

SeverinAlexB replied:
6affa91b84d420cb64814e16e5ffd8705537cb4c

SeverinAlexB commented on LSPS1/LSPS1.yaml:
```diff
@@ -0,0 +1,430 @@
+          - type: object
+            properties:
+              min_satoshi:
+                type: string
```
We should discuss this on the call. I tend to agree with you.

SeverinAlexB replied:
Spec defines all satoshi values as string, resolving this discussion.

SeverinAlexB commented:
> Is there a way to get a formatted openapi doc? Its pretty hard to review the openapi doc itself.

Use https://editor.swagger.io/ (see my initial PR description).

SeverinAlexB commented on LSPS1/extensions/lnurl_lud2/LSPS1.lnurl_lud2.yaml:
```diff
@@ -0,0 +1,58 @@
+openapi: 3.0.3
+info:
+  description: |
+    API documentation for channel opening request from a Lightning Network Service Provider (LSP). LNURL_LUD2.
```
I am trying to answer all your questions (also continuing from #18). Happy to be convinced to remove this extension.

### LNURL support
I don't see how we can support LNURL in general if we don't provide any additional endpoint. [LUD-02](https://github.com/lnurl/luds/blob/luds/02.md) requires a very specific callback url (<callback><?/&>k1=<k1>&remoteid=<Local LN node ID>&private=<1/0>) that needs to be supported to open the channel. Either we provide the LNURL string, or provide the required callback url with a standardized way of calling the api.

### Channel open on peer connect
I dislike the idea of opening the channel as soon as the peer connects. There is no proper request/response way to return an error code in case the channel open fails.

Let me know what you think.

ZmnSCPxj-jr replied:
As I mentioned in the meeting as well, if the channel order is paid via Lightning, it is possible to send back the error message over Lightning as well, and it is also an automatic refund mechanism.

For onchain, you need a separate mechanism to handle refunds anyway (LNURL does not provide this to you anyway), maybe that should be the failed-to-open-channel-error-reporting access point.

I do not really give much credence to manual flows where some web app accesses this API and then is a very thin UI layer on top of this API for initial users that need to use existing LNURL mechanisms.  This is supposed to be an *Application Program* Interface, not a manually-operated *User* Interface.

SeverinAlexB replied:
I believe the confusion we are having in the discussion stems from the fact that we have different requirements. Please have a look at the [motivation section](https://github.com/synonymdev/lsp-spec/blob/channel-request2/LSPS1/README.md#motivation) that I added with my recent commits.

> I just prefer the simplest approach because it means less code and less room for mistakes.

I am ready to discuss that our approach is too complex. I believe the requirement of "bitcoin onchain payments" (which is necessary to onboard new users to Lightning and to push sats to the user) makes the whole api more complicated. Maybe this can be solved and therefore removed with dual funded channels?

> As I mentioned in the meeting as well, if the channel order is paid via Lightning, it is possible to send back the error message over Lightning as well, and it is also an automatic refund mechanism.

I am happy to discuss if we should also send back the error message via Lightning.

> For onchain, you need a separate mechanism to handle refunds anyway (LNURL does not provide this to you anyway), maybe that should be the failed-to-open-channel-error-reporting access point.

Yes, it's on my todo

> I do not really give much credence to manual flows where some web app accesses this API and then is a very thin UI layer on top of this API for initial users that need to use existing LNURL mechanisms. This is supposed to be an Application Program Interface, not a manually-operated User Interface.

See the motivation section. Not sure what you mean with API vs UI.

ZmnSCPxj-jr replied:
> I am ready to discuss that our approach is too complex. I believe the requirement of "bitcoin onchain payments" (which is necessary to onboard new users to Lightning and to push sats to the user) makes the whole api more complicated. Maybe this can be solved and therefore removed with dual funded channels?

Indeed, that is the design goal of the dual-funded channel scheme; money is not even held and then refunded in the case where the server cannot (or refuses, for other reasons) serve the request.  It is the reason why we want to consider a simpler API here, as we view this as a stopgap until dual-funding / openV2 is widely deployed later.

Until then, what we need is some kind of refund negotiation for the onchain case.  I would still prefer having to give the refund address up-front, e.g. `refund_onchain_address`.  Need to specify exactly which standards servers MUST support (bech32m, bech32, base58).  For a manually-operated UI you could have the web app hold its own keys in client-side storage and provide the refund address from there, and only if the server refunds, ask the user for a refund address and then move from the web app refund address to the user refund address.  If the web app is run by the same people as the ones running the server it is the same trust requirement anyway, and if you expect the refund case to be rare then the inefficiency of -> web app refund -> user refund is not that important compared to the API complexity (everyone has to implement the API, not everyone has to write a manually-operatable web app, and we expect the refund case to be rare).

SeverinAlexB replied:
Did you have a chance to read the two different cases I described sync vs async?
We can use the same strategy for the `refund_onchain_address` as we use with the node_id. You can either provide it upfront (async case) or with the `/lsp/channels/{id}/update` endpoint at a later point (sync case). This would not increase the complexity of the API. The main complexity comes from having local_balance/onchain payments at all.

ZmnSCPxj-jr replied:
There is no motivation for why you want to provide the sync case.  Additional endpoints = more complexity.  For the webapp case, I do not see why sync case is somehow better than what I described.

ZmnSCPxj-jr replied:
> One thing we could do is add a "hey, poll now" endpoint, which is way simpler than LNURL-02 but would still accomplish removing the polling complexity from the server end.

Or just make the "LUD-02 endpoint" that endpoint, LOL.  Except LUD-02 expects that it is the endpoint that receives the client node ID, sigh.

SeverinAlexB replied:
How do you want to handle the case of a public channel marketplace like for example on Amboss? Requiring the user to add all the data at the beginning (node_id, channel sizes, refund address) will end in a UI/UX disaster. Marketplaces have a very limited amount of data not even talking about the ability to auto-generate addresses.
Additionally, LNURL requires a sync channel open.

ZmnSCPxj-jr replied:
What information does the marketplace need, anyway?  Just have a separate API for marketplace pricing information.  Have an API do one thing and do it well, not try to throw everything you think of into it, it just becomes a maintenance burden.  Focus this API on buying channels, make a separate API for marketplace info.

ZmnSCPxj-jr replied:
If you make a marketplace that pings this address... like, WTF, you would effectively DoS the LSPs everytime you queried market pricing information, because each time an LSP is contacted here you put another entry into their database, which the marketplace then does not ever claim.  Like come on.  Do the sensible thing and create a separate `GET`-able endpoint for pricing information that you put in your marketplace.  Could even reuse the caching information of HTTP for how long the pricing information would be valid,

SeverinAlexB replied:
I am not intending to push pricing information in to the `POST` request. The point I want to make is:

When a user wants to buy a channel, most (if not all) channel purchase websites use the following UX:
1. The user enters channel size, local balance.
2. The user pays the invoice.
3. The user either uses the LNURL to open the channel OR enters their pubkey manually.
4.1. If success everything is good.
4.2 If fail the user enters the `refund_onchain_address` in case onchain bitcoin was involved. Lightning refunds are done automatically.

Additionally, I would like to include this in the spec (happy to discuss if necessary):
4.3 If fail, the user may correct the `node_connection_string`. This is a major source of errors in case the user enters the node details manually.

The focus of this api is still purchasing channels. Maybe it is better to use the word "websites" and not "marketplace" to avoid unnecessary confusion.

The goal is to provide the lowest UI/UX friction as possible. As John said on the call 2 weeks ago: Yes, we could separate the API in a wallet channel purchase API and a website channel purchase API. The two apis overlap so much though that we would basically build the same thing two times.

ZmnSCPxj-jr replied:
The website case can be implemented by an LSP even without this API being exposed.  The entire point of this API specification is to have something that wallets can do directly without user intervention.  Which is the whole argument we have been having here: why are you infecting this particular API specifcation with UI concerns of a website?  Implement the two APIs and specify only the one for wallets, because nobody has the wallet use-case covered but everyone can easily build a website even without this spec.  The wallet use-case **needs** this spec.  The webstie use-case can use an internal API that is unspecified to the public.

SeverinAlexB replied:
I had a call with John yesterday and we will drop the website case including LNURL for this spec . It takes too much time to get to a common denominator. I tried to incorporate all our previous thoughts/requirements in the spec group up to when Spiral joined. We agree that it is best to focus on the wallet use case, which we know we can use within our own wallet too.

npslaney replied:
I think this makes a lot of sense, afaict, it's great that synonym has a manual flow via a website that essentially lets you "sell an LNURL link." I don't think anyone is questioning that's a valid use case.

Issue for me is that can be incorporated in multiple ways, and if you're expecting someone to come to your website with a wallet that uses LNURL and a method of payment, they don't need to incorporate an API. They already have everything they need and you've done the work to make it easy, which is great.

The real deep purpose for this API for me is so that someone can programmatically request inbound from an LSP, in which case, since they're already doing the work to incorporate this API, we can make the flow streamlined and intuitive, without an LNURL requirement. I see this API as a way for breez to talk to synonym, our LSP to work with X whatever wallet, etc. It'll be easier to get to something we want with a narrower scope at first.

SeverinAlexB replied:
I definitely emphasize with your POV and there are no bad feelings in focusing the scope.

The reason that I liked to incorporate LNURL/manual flow was the increased adoption potential. 
A developer friendly API would have increased the likelihood of adoption by a marketplace like Amboss or LN+. LNURL is widely supported in multiple wallets; so incentives would have been there to make more out of this api.

Happy to be called a dreamer in case you disagree with this approach :) But again, I got no issues with a narrower scope.

ZmnSCPxj-jr replied:
>A developer friendly API would have increased the likelihood of adoption by a marketplace like Amboss or LN+. LNURL is widely supported in multiple wallets; so incentives would have been there to make more out of this api.

I would suggest making a different API, preferably with input from Amboss or LN+ developers, so that you know what they actually want from that API.  Basically we know what *we* want from *this* API, so that is something we can give you input on, and you agreed that the reduced scope is something you could also use for your own plans as well.

For those I imagine the most important information is "how much would it cost to buy inbound liquidity from this LSP" and that is probably best as a separate `GET`-able API.

ZmnSCPxj-jr commented on LSPS1/extensions/refunds/README.md:
```diff
@@ -0,0 +1,36 @@
+
+## API
+
+The extension `refunds` adds two endpoints to automatically manage refunds.
```
Refunds are not strictly needed if LSP is only accepting Lightning payments, as refusing to release the preimage is automatically a refund.  But if you are going to use onchain payments at all, it needs refunds unless you want to do onchain contracting of some kind.

SeverinAlexB replied:
Agree. I worked on the spec the last days and I incorporated HOLD invoices (see new commits). Onchain needs another refund mechanism (todo).

ZmnSCPxj-jr replied:
Why hold it though?  Just refuse it outright and give the error reason in the new fat error mechanism.

SeverinAlexB replied:
Normal invoices (cln, lnd, not idea about LDK) release the preimage instantly when the payment arrives. You need to hold the payment for the duration of the channel open process at least.

1. Payment arrives at the LSP.
2. LSP holds payment.
3. LSP tries to open the channel.
4. LSP resolves/rejects payment depending on the channel open success.

ZmnSCPxj-jr replied:
What I meant is that as soon as LSP *allocates* money for it, claim the payment.  If LSP cannot allocate money, reject it.  No need to wait for network connection or channel opening or confirmation of channel.

This can be implemented in Core Lightning via `htlc_accepted` plugin hook so that the plugin can wait for money allocation procedure, and `signinvoice` command to create a non-automatic invoice.  I assume LND has similar since peerswap is implemented for it and peerswap would (probably?) use something similar, or maybe not.  LDK still needs some hooks on payment release I think.

SeverinAlexB replied:
I need to specify what error the LSP should set in case it doesnt have funds. The LSP SHOULD already throw this error at the order creation call `POST /lsp/channels`. It MUST set the error after the channel open failed.

I am unsure what you mean with the "money allocation" step. In the case you open every channel with it's own transaction, the money is allocated right before the funding transaction is published. The payment SHOULD be claimed after the funding transaction is published. I don't see a DDoS vector here.

Are you referring to the batched channel open case? Here I can see that the steps `money allocation` and `funding transaction published` could be different. I have not thought through this case completely yet though.

ZmnSCPxj-jr replied:
>The LSP SHOULD already throw this error at the order creation call `POST /lsp/channels`

If you reserve funds before getting paid, you are left vulnerable to someone spamming `POST /lsp/channels` on you and preventing you from allocating funds for other clients.  You have to reserve funds at the point you get paid, not at `POST` time.

>I am unsure what you mean with the "money allocation" step

Situation is this:

* LSP has 1 BTC.
* Client A wants 1 BTC of capacity
* Client B wants 1 BTC of capacity
* Client A makes a `POST` request.
* Client B makes a `POST` request.
  *  If LSP errors at this point because LSP already allocated 1.0 BTC to client A, client A can simply not pay the invoice, thus keeping the 1 BTC reserved for client A, and any future client Bs are unable to acquire service, i.e. denial of service.
* Client B tries to pay the invoice.  LSP claims the payment and reserves 1.0 BTC to client B, waiting for client B to connect or ping the LNURL (why tho) or whatever.
* Client A tries to pay the invoice.
  * This is the proper place to report the error and fail to claim the invoice payment, and is what I am referring to.

Basically you have to allocate funds, at some point, and that allocation can fail because nobody has infinite Bitcoins, it is literally just 21 million.  If you error early, before getting paid, as noted above client A can just reserve your funds.

Allocation need not be tied to acutally publishing funding transaction.  Presumably the LSP knows the value of every UTXO it controls, and can easily sum them up and check if the reserved funds is less than or equal to that value.

ZmnSCPxj-jr commented on LSPS1/extensions/refunds/LSPS1.refunds.yaml:
```diff
@@ -0,0 +1,154 @@
+  title: "LSP Spec 1"
+tags:
+- name: "Refund"
+  description: "Extension: `refunds`"
```
W/hy is this an extension rather than part of base?

SeverinAlexB replied:
Refunds are still a todo. Ignore all extensions for the moment.

ZmnSCPxj-jr replied:
Cannot really do onchain without refunds.  Yes, you can add all sorts of stuff to make sure you "never" run out of money to service incoming requests, but accidents happen and refund mechanisms are necessary.  Either cut out onchain completely from base and put onchain+refund on an extension, or have them both in base.  I suggest have them both in base.

SeverinAlexB replied:
Again, this is on my todo list. Will have a look at it next week.

SeverinAlexB replied:
Please have a look at my update in the PR and let me know if you agree with it.

ZmnSCPxj-jr commented on LSPS1/README.md:
```diff
@@ -0,0 +1,386 @@
+
+This section describes the payment object returned by `POST /lsp/channel` and `GET /lsp/channel/{id}`. The user MUST pay the `lightning_invoice` OR the `btc_address`. Using both methods MAY lead to the loss of funds.
+
+> **Rationel** Onchain Payments are required for payments with higher amounts, especially to push local_balance_satoshi to the user. Onchain payments are also useful to onboard new user to Lightining. Lightning payments are the preferred way to do payments because they are quick and easily refundable.
```
s/Rationel/Rationale/

ZmnSCPxj-jr commented on LSPS1/README.md:
```diff
@@ -0,0 +1,386 @@
+
+### Channel sides
+
+Channel sides are seen from the user point of view. `local_balance` are the funds on the user side. `remote_balance` are the funds on the LSP side.
```
Maybe better to be explicit and say `client_balance` and `lsp_balance` instead?

SeverinAlexB replied:
Done.

SeverinAlexB commented on LSPS1/README.md:
```diff
@@ -0,0 +1,386 @@
+    - `remote_balance_satoshi` MUST be 1 or greater. MUST be below or equal `base_api.max_remote_balance_satoshi`.
+    - `local_balance_satoshi` MUST be 0 or greater. MUST be below or equal `base_api.max_local_balance_satoshi`. Todo: Rejection error message.
+    - `onchain_fee_rate` MUST be 1 or higher. The LSP may increase this value depending on the onchain fee environment. MAY be unspecified, the LSP will determine the fee rate. 
+    - `channel_expiry_weeks` MUST be 1 or greater. MUST be below or equal `base_api.max_channel_expiry_weeks`.
```
Hmmm, this was actually never really discussed as far as I know. It was just assumed by all parties. We should discuss this on our next call. I am open for `days` or `hours`.

SeverinAlexB replied:
We decided to use `blocks` as the measure of time.

SeverinAlexB commented:
I updated the PR. Some highlights:

- Change naming of remote/local to user/lsp. For example, user_balance and lsp_balance.
- Removed sync flow for websites.
- Simplified peer connection. Only let user open the connection.
- Added node connection string definition.
- Remove all extensions (jit, lnurl, refund). JIT is still an open question though.
- Added onchain refund flow.
- Added more open questions at the bottom of the doc.

SeverinAlexB commented on LSPS1/README.md:
```diff
@@ -0,0 +1,386 @@
+**User**
+
+If `node_connection_string_or_pubkey` contains a full connection string:
+- SHOULD open a peer connection before the channel is being opened.
```
Please have a look at my update in the PR and let me know if you agree with it.

ZmnSCPxj-jr commented:
For JIT channels, I would suggest making a separate spec.  Basically, this LSPS1 is for pre-buying inbound liquidity, then specify a different spec for JIT channels.

ZmnSCPxj-jr commented on LSPS1/README.md:
````diff
@@ -0,0 +1,409 @@
+```json
+"options": {
+    "base_api": {
+        "version": 1,
````
Maybe better for this query to return a *set* of supported versions, e.g. `[1, 2]`, or for current `[1]`.  In the `POST` request the client indicates a specific version number.

ZmnSCPxj-jr commented on LSPS1/README.md:
```diff
@@ -0,0 +1,409 @@
+- MAY implement RBF for channel opening.
+  - In that case, the client will see multiple parallel channel open requests, each one being a different version of the channel opening transaction.
+
+// Todo: How does RBF work with 0conf?
```
It does not :)

ZmnSCPxj-jr commented on LSPS1/README.md:
```diff
@@ -0,0 +1,409 @@
+
+The user MAY abort the flow here.
+
+### 2. Create Order
```
You started numbering this with `2.` but the previous subsection "API Information" was not numbered.

ZmnSCPxj-jr commented on LSPS1/README.md:
````diff
@@ -0,0 +1,409 @@
+**Example request body**
+
+```json
+{
````
The `GET` response returns a version, but this `POST` request does not include a version field?

SeverinAlexB commented:
- Changed `channel_expiry_weeks` to `channel_expiry_blocks` as discussed in the meeting.
- Use the same version array as proposed on LSPS3. This allows for one api to support multiple versions.
- Changed `onchain_fee_rate` to `confirms_within_blocks`. This makes it more clear on how fast the channel should confirm onchain. It limits the amount of time an LSP can wait to batch channel opens and gives the user clear expectations on when the channel should be open. See transcription of our last meeting for more information.
- Added `minimum_depth` option that indicates how many block confirmations the LSP requires. `0` means the LSP supports 0conf.
- Added `supports_zero_channel_reserve` option to indicate if the LSP allows 0reserve.

`minimum_depth` and `supports_zero_channel_reserve` options are my proposal following the discussion around 0conf and 0reserve support. It clearly indicates if these non-standard methods are supported.

yaslama commented on LSPS1/LSPS1.yaml:
```diff
@@ -0,0 +1,489 @@
+  
+    post:
+      tags:
+      - "Channel Request"
```
If we want to support HTTP, gRPC and onion messages, it's better to have different paths for the information and request a channel. What about  GET /lsp/channelinformation and POST /lsp/channel. I am open to any other suggestion

SeverinAlexB replied:
I am not 100% sure yet if we should implement the API in several formats as this doubles the implementation work for LSPs. One single format would make sure everything is interoperable.

Nonetheless my concerns aside, if we go ahead with this I propose the following format:

1. General formats and requirements.
2. HTTP requirements.
3. gRPC requirements.

HTTP is supported by the OpenApi doc and gRPC should be supported by a Interface Definition Language (IDL) doc.

Let me make an example for the `GET /lsp/channels` endpoint:

## api_information_endpoint

The api_information_endpoint is the entrypoint for each client using the api. It lists the versions of the api and all options in a dictionary.

The user MUST call api_information_endpoint to read

- the `versions` of the api and therefore prove compatibility.
- `options` properties to determine the api boundaries.

Example api_information_endpoint response: 

```JSON
{
  "versions": [2],
  "website": "http://example.com/contact",
  "options": {
      "minimum_depth": 0,
      "supports_zero_channel_reserve": true,
      "max_user_balance_satoshi": "0",
      "max_lsp_balance_satoshi": "100000000",
      "min_required_onchain_satoshi": null,
      "max_channel_expiry_blocks": 20160
  }
}
```

> **HTTP** Method: GET; Path: /lsp/channels; Code 200
> Any other http details for this api_information_endpoint go here.

> **gRPC** Service: lsps1; Rpc: apiInfo
> Any other gRPC details for this api_information_endpoint go here.

SeverinAlexB replied:
Gonna close this due to LSPS0.

SeverinAlexB commented on LSPS1/README.md:
```diff
@@ -0,0 +1,392 @@
+
+Word definition:
+
+* `<node>` MUST be a node id (pubkey). For example: "0200000000a3eff613189ca6c4070c89206ad658e286751eca1f29262948247a5f".
```
I added references. nodeid ref just points to BOLT8 as it is not defined better anywhere else.

SeverinAlexB commented on LSPS1/README.md:
```diff
@@ -0,0 +1,392 @@
+Node connection strings like `lsp_connection_string` MUST match one of the following grammars:
+
+* `<node> '@' <ip> `:` <port>`, with the last `:` after `@` considered a separator for the port number.
+* `<node> '@' '[' <ip> ']' ':' <port>`, with any `:` inside `<ip>` considered part of the IP address.
```
Sure but I guess the best approach here is to add them in a spec update later. Or what are your thoughts exactly?

SeverinAlexB commented on LSPS1/README.md:
```diff
@@ -0,0 +1,392 @@
+    - `user_balance_satoshi` MUST be 0 or greater. MUST be below or equal `base_api.max_user_balance_satoshi`. Todo: Rejection error message.
+    - `confirms_within_blocks` MUST be 0 or higher.
+    - `channel_expiry_blocks` MUST be 1 or greater. MUST be below or equal `base_api.max_channel_expiry_blocks`.
+    - `coupon_code` MUST be a string or null.
```
My rationale: `coupon_code` can be used as a discount code and potentially as a auth mechanism (to be further discussed?). Do you see other use cases?

SeverinAlexB commented:
Summary of the changes:
- Removed everything that had to do with connection_strings. They are not needed already.
- Removed the peer connection section. With JSON RPC, we assume we have a peer connection anyway.
- Renamed `user` to `client` to make it similar to `LSPS0`.
- Turned all HTTP requests (aka `POST /lsp/channel`) into jsonrpc method names like `lsps1.create_order`.
- Added Error section to all defined methods. Heavily inspired by Zmn's proposal in #3.
- Removed the OpenAPI doc as it is incompatible with jsonrpc. I moved all the field descriptions to the readme.
  - This increased the size of the doc significantly but made it more readable IMO.
- Added more options to better define what the LSP is willing to sell:
  - min_client_balance_satoshi
  - max_client_balance_satoshi
  - min_lsp_balance_satoshi
  - max_lsp_balance_satoshi
  - min_channel_balance_satoshi
  - max_channel_balance_satoshi
- Added `api_version` to the order object.
- Make it explicit that if a `coupon_code` is invalid, a error will be returned.
- Added `Idempotent Yes/No` to every request to indicate it shouldn't have side effects.
- Described the Channel object in further detail.

SeverinAlexB commented:
# Summary of changes

Added sections and requirements for 0conf risk management. As described in this [article](https://medium.com/@akumaigorodski/instant-channels-enable-safe-lightning-payments-with-unconfirmed-funding-8d640defa183), if we are sure of the onchain fee+client_balance payment, the LSP has zero risks opening a 0conf channel. Therefore the spec focuses fully on securing the onchain payment and NOT on the 0conf channel open.

## Onchain payment

I introduced 2 new fields:
- `onchain_block_confirmations_required` Number of block confirmations required for the LSP to accept the onchain payment as confirmed. 0 would indicate 0conf payments.
- `minimum_fee_for_0conf` Minimum fee that the LSP requires to accept the onchain payment as 0conf. This is only necessary in case `onchain_block_confirmations_required` is 0 of course.

These are the two tools the LSP can work with to mitigate the onchain payment risk. Important to note here: Both fields are "best-guesses". If the LSP doesn't feel comfortable in these values anymore when the client actually pays, the LSP MAY wait longer for confirmation. The whole approach to risk management can be described as: The LSP tries to estimate the risk-mitigating values as good as possible but may always back off in case it is not confident in them anymore (for example during a fee spike).

## Other changes

- Renamed `coupon_code` to `token`.
- Referenced LSPS0 common schema data types.


At this point, I think the spec is basically complete. Please review it so we can potentially merge it in some weeks.

SeverinAlexB commented on LSPS1/README.md:
```diff
@@ -0,0 +1,503 @@
+      "minimum_depth_channel": 0,
+      "minimum_onchain_payment_confirmations": 1,
+      "supports_zero_channel_reserve": true,
+      "min_required_onchain_sat": null,
```
Will change to `min_onchain_payment_size_sat`.

>Also, I'm not quite sure why this field is necessary, couldn't the LSP simply leave out the onchain funding address in case the client creates an order smaller than this amount?

Some thoughts about `info()`:
- This data should not change that often. Explorer like Amboss should be able to list it on their website.
- Reading `min_onchain_payment_size_sat` will prevent users from creating orders. This reduces the servers requirements to watch onchain addresses and create lightning invoices.

I guess I need to add this rationale to the spec.

SeverinAlexB commented on LSPS1/README.md:
```diff
@@ -0,0 +1,503 @@
+      "max_client_balance_sat": "100000000",
+      "min_lsp_balance_sat": "0",
+      "max_lsp_balance_sat": "100000000",
+      "min_channel_balance_sat": "50000",
```
- Why are they redundant in your opinion?
- Will rename `min_channel_balance_sat` to `min_channel_size_sat`.

SeverinAlexB replied:
Example why this doesnt work:

Blocktank has a (legal) channel limit of $1000. User can ask for a $1000 `lsp_balance_sat` channel or $500 `lsp_balance_sat` + $500 `client_balance_sat`.

max_lsp_balance_sat=$1000
max_client_balance_sat=$500 (half of max_lsp_balance_sat).
max_channel_size_sat=$1000

`max_lsp_balance_sat` + `max_client_balance_sat` != `max_channel_size_sat` in this case.

SeverinAlexB commented on LSPS1/README.md:
```diff
@@ -0,0 +1,503 @@
+    - `lsp_balance_sat` *LSPS0.sat* How many satoshi the LSP will provide on their side.
+      - MUST be 1 or greater. 
+      - MUST be equal or below `base_api.max_lsp_balance_sat`.
+      - MUST be equal or greater `base_api.min_lsp_balance_sat`.
```
I assume only for `lsp_balance_sat`? If we also require this for `client_balance_sat` then we make "push_balance_to_client" mandatory which could lead to legal issues.

SeverinAlexB replied:
Actually, LND only supports `client_balance_sat < lsp_balance_sat`. We could solve this by introducing a new option called `max_client_balance_ratio_ppm`.

- `max_client_balance_ratio_ppm <LSPS0.ppm>` Maximum size of `client_balance_sat` compared to `lsp_balance_sat`. 
  - Formula: `max_client_balance_sat = lsp_balance_sat * (max_client_balance_ratio_ppm / 1000000)`
  - LND would be 500,000 (50%).
  - Regular channel reserve would be 10,000 (1%).
  - 0reverse would be 0 (0%).
  
What do you think?

SeverinAlexB commented on LSPS1/README.md:
```diff
@@ -0,0 +1,503 @@
+
+Any error MAY include a `message` field that is a human-readable string. The `message` field is intended to provide a short description of the error. The `message` field is not intended to be parsed by the client or shown to the user. It is intended as a developer hint to help the client developer debug the error.
+
+## Overview
```
Refunds are explained in case the channel open failed. This is the most important failure scenario IMO. I don't see any other scenario that is important enough to include here. Please point to one if you think it I should include it.

SeverinAlexB commented on LSPS1/README.md:
```diff
@@ -0,0 +1,503 @@
+2. Client calls `lsps1.create_order` to create an order.
+3. Client pays the order either onchain or offchain.
+4. LSP opens the channel as soon as they payment is confirmed.
+5. LSP refunds the client in case the channel open failed.
```
I separated it more.

SeverinAlexB commented on LSPS1/README.md:
```diff
@@ -0,0 +1,503 @@
+    - MAY be 0 to allow 0conf payments.
+    - MUST be 0 or greater.
+  - `supports_zero_channel_reserve` *boolean* Indicates if the LSP supports [zeroreserve](https://github.com/ElementsProject/lightning/pull/5315).
+  - `min_required_onchain_sat` *LSPS0.sat or null* Indicates the minimum amount of satoshi (`order_total_sat` see below) that is required for the LSP to accept a payment onchain.
```
I tried multiple formats. Making the types bold doesn't make it better (I tried). The best I found was the following:
- `field_name <type>` Description

Let me know if you still think it is insufficient.

SeverinAlexB commented on LSPS1/README.md:
```diff
@@ -0,0 +1,503 @@
+    - `refund_onchain_address` *LSPS0.onchain_address* Address where the LSP will send the funds if the order fails.
+      - Client MAY omit this field.
+      - LSP MUST disable onchain payments if the client did omit this field.
+- `announceChannel` *boolean* If the channel should be announced to the network (also known as public).
```
I made everything flat.

SeverinAlexB commented on LSPS1/README.md:
````diff
@@ -0,0 +1,503 @@
+}
+```
+
+- `id` *string* Id of this specific order.
````
Changed to `order_id`.

SeverinAlexB commented on LSPS1/README.md:
```diff
@@ -0,0 +1,503 @@
+**LSP**
+
+- MUST change the payment state to `HOLD` when the payment arrived.
+- MUST set open.state to `PENDING`.
```
We have multiple states in this spec.
- `payment.state`
- `channel.state`
- `state` which I renamed to `order_state` to distinguish them better.

SeverinAlexB commented on LSPS1/README.md:
```diff
@@ -0,0 +1,503 @@
+  - In this case, the LSP MUST still ensure that the funding transaction gets confirmed after a maximum of `confirms_within_blocks` after the client payment completed.
+
+
+### 5 Channel Object
```
I did move it up. Not sure if you meant that exactly.

SeverinAlexB commented:
I fixed basically all comments from @tnull. See commit messages.
Some highlights:
- Renamed `state` to `order_state` to distinguish them better from `payment.state` and `channel.state`.
- Found some invalid `order_state` changes and fixed them.
- Added `SHOULD limit changes in `lsps1.info` to once/day. See rationale in the spec for more info.
- Changed how types are shown
  - From: `property` *type* Description
  - To: `property <type>` Description
- Fixed lots of typos.

I left some comments open above that still need inputs.

SeverinAlexB commented on LSPS1/README.md:
````diff
@@ -0,0 +1,392 @@
+}
+```
+
+- `api_version` MUST be `2`. MUST match one of the versions listed by the API.
````
Change to `1`.

SeverinAlexB commented on LSPS1/README.md:
```diff
@@ -0,0 +1,488 @@
+  - MUST be below or equal `base_api.max_channel_expiry_blocks`.
+- `token <string>` Field for arbitrary data like a coupon code or a authentication token.
+  - Client MAY omit this field.
+  - Maximum length is 512 characters.
```
Done

SeverinAlexB commented:
Some minor changes:
- Added trust model
- Changed API version from 2 to 1
- Put null into `null`.
- Removed token max_length.

I hope I now caught up on all comments from @tnull that I missed.

SeverinAlexB commented:
Added Anchor links. It's a bit of a mess though as the `<>` characters outside of "`" ticks hide content in some cases. If anybody has a better idea on how to display the types visually, let me know.

@tnull

SeverinAlexB commented:
>Oh, could you give an example where this is the case? Looked alright from a quick scroll-through for me. Also, might be nice to rebase this PR for now on https://github.com/BitcoinAndLightningLayerSpecs/lsp/pull/42 so that the links actually work.

Github markdown rendering: 
- `Array<uint16>` will render to simple Array<uint16>.

Screenshot of what I typed:
![CleanShot 2023-06-16 at 15 12 07@2x](https://github.com/BitcoinAndLightningLayerSpecs/lsp/assets/8782386/c7316177-b1a1-4b38-bc78-d98b9a41d90f)

So content between <> just disappears. If I put the types in `` then the links don't render anymore.

Github markdown rendering:
- [Link](www.example.com) will be `[Link](www.example.com)`.

Screenshot of what I typed:
![CleanShot 2023-06-16 at 15 14 41@2x](https://github.com/BitcoinAndLightningLayerSpecs/lsp/assets/8782386/1a424355-10f5-468b-83cc-3829883c9966)


So I had to put the types into \`\` sometimes and outside of \`\` the other time. It's a mess.
<img width="326" alt="CleanShot 2023-06-16 at 15 16 07@2x" src="https://github.com/BitcoinAndLightningLayerSpecs/lsp/assets/8782386/aa37f677-5819-492c-97c3-cc1b683bbf2d">

<img width="447" alt="CleanShot 2023-06-16 at 15 17 54@2x" src="https://github.com/BitcoinAndLightningLayerSpecs/lsp/assets/8782386/e0c38d09-c86d-4d3d-ac13-76da88c31336">

---
And yes, I will rebase.

SeverinAlexB commented:
> > Github markdown rendering:
> > 
> > * `Array<uint16>` will render to simple Array.
> > 
> > Screenshot of what I typed: ![CleanShot 2023-06-16 at 15 12 07@2x](https://user-images.githubusercontent.com/8782386/246343874-c7316177-b1a1-4b38-bc78-d98b9a41d90f.png)
> > So content between <> just disappears. If I put the types in `` then the links don't render anymore.
> > Github markdown rendering:
> > 
> > * [Link](www.example.com) will be `[Link](www.example.com)`.
> > 
> > Screenshot of what I typed: ![CleanShot 2023-06-16 at 15 14 41@2x](https://user-images.githubusercontent.com/8782386/246344493-1a424355-10f5-468b-83cc-3829883c9966.png)
> > So I had to put the types into `sometimes and outside of` the other time. It's a mess. <img alt="CleanShot 2023-06-16 at 15 16 07@2x" width="326" src="https://user-images.githubusercontent.com/8782386/246345102-aa37f677-5819-492c-97c3-cc1b683bbf2d.png">
> > <img alt="CleanShot 2023-06-16 at 15 17 54@2x" width="447" src="https://user-images.githubusercontent.com/8782386/246345193-e0c38d09-c86d-4d3d-ac13-76da88c31336.png">
> > And yes, I will rebase.
> 
> Mh, but usually [`Array<uint16>`](https://www.example.com) should work? That is
> 
> ```
> [`Array<uint16>`](https://www.example.com)
> ```

Yes, but what link do you want to put behind the array? We have LSPS0 data types and we have regular data types like the array.

SeverinAlexB commented on LSPS1/README.md:
```diff
@@ -0,0 +1,498 @@
+> **Rationale Client rejected** LSPs can reject a client for example for misbehaviour. LSPs can reject a node on two levels: Prevent a peer connection OR disable order creation. Preventing a peer connection might not work in case you still want to allow other functions to keep working, for example an existing channel.
+
+
+### 2.1 lsps1.get_order 
```
I mean, as we already use `snake_case` everywhere, why should we us a different naming schema in method names?
@ZmnSCPxj-jr

ErikDeSmedt commented on LSPS1/README.md:
```diff
@@ -0,0 +1,498 @@
+  - `minimum_channel_confirmations <unit8>` Minimum number of block confirmations before the LSP accepts a channel as confirmed and sends [channel_ready](https://github.com/lightning/bolts/blob/master/02-peer-protocol.md#the-channel_ready-message) (previously `funding_locked`).
+    - MAY be 0 to allow 0conf channels.
+    - MUST be 0 or greater.
+  - `minimum_onchain_payment_confirmations <unit8>` Minimum number of block confirmations before the LSP accepts an on-chain payment as confirmed. This is a lower bound. The LSP MAY increase this value by responding with a different value in `lsps1.create_order.onchain_block_confirmations_required` depending on the size of the channels and risk management.
```
Typo, I assume this should be `uint8` instead of `unit8`
